### ISSUE TITLE
Missing semantics: FXSAVE/FXRSTOR

### ISSUE THREAD

**Issue by fvrmatteo** [JonathanSalwan/Triton#875]
The following instructions semantics are missing from Triton:
- **fxsave**: https://www.felixcloutier.com/x86/fxsave
- **fxrstor**: https://www.felixcloutier.com/x86/fxrstor

Supporting them would require to:
1. Have a reliable way to distinguish between the different execution modes (e.g. non-64-bit operation modes or 64-bit sub-mode of the IA-32e mode).
2. Implement the missing FPU registers (with the possibility to symbolise, taint and concretise them).

Other binary lifters (e.g. Remill, RetDec, Angr) seem to be missing these instructions or implementing them as inline assembly calls.

What would it be the best way to support the store and load of the FPU, MMX and SSE state?

This is likely tied to https://github.com/JonathanSalwan/Triton/issues/326, but I think with enough effort it's possible to implement the semantic after solving the requirements. The accessed values are not used in any meaningful way apart from being read/written from/to a register or memory location.

**Comment by XVilka** [comment]
Should be this closed then? Since the PR in question was merged already.

**Comment by JonathanSalwan** [comment]
@XVilka nop. I've merged to fix some issues but we are currently working on it with @fvrmatteo on his repo (https://github.com/fvrmatteo/Triton/commits/dev-v0.8). Once it's done, I will merge into `dev-arm32` which will be merged in `dev-v0.8` in a near future.

**Comment by fvrmatteo** [comment]
Just to track down what is missing:

1. implementing the check on the EFER->FFXSR bit flag to execute the fast FXSAVE(64)/FXRSTOR(64)
2. if possible moving the MMX registers from "128 bit + mask" to plain "80 bit Boost type"
3. implementing the exception when the address is not 16 byte aligned (need to wait for Triton's support)

I accept suggestions for the implementation of the 80 bit Boost type using the plain Boost library (no external dependencies). As of now the only solution I found would be to include the headers of this [Boost.LargeInt](https://github.com/mjtryhorn/Boost.LargeInt) library into the includes folder.

**Comment by fvrmatteo** [comment]
Ok managed to have 80-bit values with plain Boost:
```
typedef boost::multiprecision::number<boost::multiprecision::cpp_int_backend<80, 80, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void> > uint80;
```

@JonathanSalwan Would you accept the usage of this type for the MMX registers if we can confirm that everything works? :)

**Comment by JonathanSalwan** [comment]
Well, it depends. Where this type will be used? In Triton, everything is in 512 bit and values like 128 and 256 are mainly used far casting. With this new type can we still apply this: `value.convert_to<triton::uint80>()` where `value` is a `boost::multiprecision::uint512_t` ?

**Comment by fvrmatteo** [comment]
Yes the type is in fact defined in the same way of the other Boost types used by Triton:
```

**Comment by JonathanSalwan** [comment]
Great then :)

**Comment by fvrmatteo** [comment]
The implementation is complete, it's using the 80 bit defined type, but I had to rewrite the **getConcreteRegisterValue** and **setConcreteRegisterValue** functions to use **std::memcpy** to avoid wrong type punning. I can confirm the call to **std::memcpy** is detected as a pattern by the compiler, inlined and correct TBAA information are generated. In some cases the **std::memcpy** is not inlined, but it was the same with the old code when dealing with 128/256/512 bits variables.

The semantic is now handling checks on CPL=0 and the FFXSR bit of the EFER register, so I had to add support for new flags registers (all the bits of EFER). The only thing still missing is the generation of the exception in case an unaligned pointer is detected.

The comparison of FXSAVE/FXSAVE64 with Unicorn differs only on the value of some reserved bytes in the 512 bytes layout. The FXRSTOR/FXRSTO64 match completely.

I think I'll need to open a new PR to review, improve and merge the code, I just need to know on which branch it's best to work on.

The implemented semantics are the following:
- [FXSAVE](https://github.com/fvrmatteo/Triton/blob/dev-v0.8/src/libtriton/arch/x86/x86Semantics.cpp#L6152)
- [FXSAVE64](https://github.com/fvrmatteo/Triton/blob/dev-v0.8/src/libtriton/arch/x86/x86Semantics.cpp#L6455)
- [FXRSTOR](https://github.com/fvrmatteo/Triton/blob/dev-v0.8/src/libtriton/arch/x86/x86Semantics.cpp#L5241)
- [FXRSTOR64](https://github.com/fvrmatteo/Triton/blob/dev-v0.8/src/libtriton/arch/x86/x86Semantics.cpp#L5706)

**Comment by JonathanSalwan** [comment]
Thanks for this contribution Matteo! I've merged what we did on `dev-arm32` into `dev-v0.8`, so you can now rebase on `dev-v0.8`. Note that I've pushed something that may imply some conflicts with your code. See #877. Once our MR is on, I will start review your code and merge it in `dev-v0.8`.

**Comment by JonathanSalwan** [comment]
We plan to release the `dev-v0.8` this week or next week. So we will merge this code into `dev-v0.9` so that we can test it a bit before release.

**Comment by JonathanSalwan** [comment]
Merged in `dev-fpu`.  I will merge it to `dev-v0.9` once unittests are done.

**Comment by mrexodia** [comment]
I found https://github.com/JonathanSalwan/Triton/blob/af51f4b75fbee008db6f1160d135204316aa5ccb/src/testers/x86/unicorn_test_x86.py, it seems the DBI tests mentioned [here](https://github.com/JonathanSalwan/Triton/pull/882#issuecomment-708611078) seem to be removed.

The issue is that qemu/unicorn is emulating these instructions incorrectly, but I can make a test anyway. Perhaps I can create a test application, record inputs/outputs and just compare against that?

**Comment by JonathanSalwan** [comment]
> @JonathanSalwan Could you give a bit of guidance on how to execute the unit tests and which ones need to be added?

`make check` does all tests. There are actually 3 kind of tests:

* [1] Python unit tests which can be tested with the following command (**OSX, Linux, Windows**)
  * `$ python -m unittest discover src/testers/unittests`
* [2] Python standalone scripts that use unicorn to check semantics, for example: (**Linux only**)
  * `$ python src/testers/aarch64/unicorn_test_aarch64.py`
* [3] C++ test which is compiled by `cmake` and can be tested executing the following program (I made this program to test things that can not be tested with Python bindings) (**Linux, OSX only**)
  * `build/src/examples/cpp/ctest_api`. 

When running `make check` it executes those 3 above checks. However, if you want to test specifics things, I suggest to edit or add a file in `src/testers/unittests` because this is the only test that is performed on Linux, OSX and Windows.

Back to testing the FPU, I think the best way is to emulate instructions with Triton and then compare results with hardcoded values. Thus you can add this in `src/testers/unittests` without involving Qemu/Unicorn.

**Comment by JonathanSalwan** [comment]
I will work on it this week. Rebasing from dev-v0.9 will be boring so I will commit changes with @fvrmatteo's name. @fvrmatteo is it ok for you?

**Comment by JonathanSalwan** [comment]
I think after this we can release the dev-0.9

**Comment by fvrmatteo** [comment]
Sounds ok to me @JonathanSalwan! Looking forward to seeing how you'll end up testing this, so maybe for the next time I'll do the same. Thanks for getting back to this :)

**Comment by mrexodia** [comment]
Why not just merge the branch and leave things as they are? ^^ (After adding/merging tests of course)

**Comment by JonathanSalwan** [comment]
We can't merge, there is too much conflicts.

**Comment by JonathanSalwan** [comment]
But if fixing conflicts do not afraid you I can let you doing this :)

**Comment by JonathanSalwan** [comment]
I think the easiest and lazy way to rebase it is to copy / past [all lines](https://github.com/JonathanSalwan/Triton/compare/dev-v0.9...dev-fpu) in a new commit.

**Comment by mrexodia** [comment]
That’s exactly what a merge does haha. It doesn’t really matter for me though, you can also do another commit with Matteo’s name and email. I can take a look at fixing the conflicts though np

**Comment by mrexodia** [comment]
[0001-Improve-tritonConfig.cmake.zip](https://github.com/JonathanSalwan/Triton/files/8018389/0001-Improve-tritonConfig.cmake.zip)

This patch should merge `dev-0.9` into `dev-fpu` and allow you to continue work without conflicts. You could also merge `dev-fpu` into `dev-0.9` and work on top, there are only 2 conflicts that instantly get resolved for me with kdiff3 ^^

```
git checkout dev-fpu
git am 0001-Improve-tritonConfig.cmake.patch
```

**Comment by mrexodia** [comment]
Not sure how to share it other than this 😅

```
git remote add llvmparty https://github.com/LLVMParty/Triton/commit
git fetch llvmparty dev-v0.9
git checkout dev-v0.9
git reset --hard a10e3c6e350bab6ccf40d3873d71405107dcb1f9
git remote remove llvmparty
```

**Comment by JonathanSalwan** [comment]
Thanks, i will work on it this evening.

**Comment by JonathanSalwan** [comment]
@mrexodia, @fvrmatteo, everything are merged into `dev-v0.9` with a [minimum](https://github.com/JonathanSalwan/Triton/blob/fdb417815b344dea57875b99795ab761cfb7f1a6/src/testers/unittests/test_semantics.py#L457-L565) of testing. If you want to improve this file to add more test around fpu, feel free!
